openslides_backend/action/motion_category/number_motions.py

```diff
@@ -0,0 +1,225 @@
+from ..base import ActionPayload
+from ..default_schema import DefaultSchema
+from ..generics import UpdateAction
+from ..register import register_action
+
+
+@register_action("motion_category.number_motions")
+class MotionCategoryNumberMotions(UpdateAction):
+    """
+    Action to number motions for a category.
+    """
+
+    model = Motion()
+    schema = DefaultSchema(MotionCategory()).get_default_schema(
+        title="Number motions of a motion category.",
+        description="An object containing an array of main category id.",
+        required_properties=["id"],
+    )
+
+    def get_updated_instances(self, payload: ActionPayload) -> ActionPayload:
+        for instance in payload:
+            self.init_memory(instance["id"])
+
+            affected_categories = self.get_affected_categories(instance["id"])
+            affected_motions = self.get_affected_motions(affected_categories)
+            non_affected_motions = [
+                id_ for id_ in self.mem_motions if id_ not in affected_motions
+            ]
+            existing_numbers = [
+                self.mem_motions[id_].get("number") for id_ in non_affected_motions
+            ]
+
+            # check for missing lead_motion_ids in affected_motions.
+            for motion_id in affected_motions:
+                lead_motion_id = self.get_lead_motion_id(motion_id)
+                if lead_motion_id and lead_motion_id not in affected_motions:
+                    raise ActionException(
+                        f'Amendment "{motion_id}" cannot be numbered, because it\'s lead motion ({lead_motion_id}) is not in category {instance["id"]} or any subcategory.'
+                    )
+
+            # generate number_value_map with the number_values for motion_ids.
+            main_counter = 1
+            lead_motions_counter: Dict[int, int] = dict()
+            number_value_map: Dict[int, int] = dict()
+            for motion_id in affected_motions:
+                lead = self.get_lead_motion_id(motion_id)
+                if lead:
+                    if lead in lead_motions_counter:
+                        number_value_map[motion_id] = lead_motions_counter[lead]
+                        lead_motions_counter[lead] += 1
+                    else:
+                        number_value_map[motion_id] = 1
+                        lead_motions_counter[lead] = 2
+                else:
+                    number_value_map[motion_id] = main_counter
+                    main_counter += 1
+
+            for motion_id in affected_motions:
+                number = self.get_number(motion_id, number_value_map)
+                if number in existing_numbers:
+                    raise ActionException(
+                        f'Numbering aborted because the motion identifier "{number}" already exists.'
+                    )
+
+                yield {
+                    "id": motion_id,
+                    "number": number,
```

Comment:
the `number_value` should also be set. Maybe PR #336 should be merged first.

Reply:
Added number_value to the instance.